modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,197 @@
+
+                weightedDistribution <- EitherT.fromEither(weight(distribution))
+
+                transactions <- EitherT.fromEither {
+                  weightedDistribution.toNel.toList.traverse {
+                    case (address, reward) =>
+                      PosLong
+                        .from(reward.value)
+                        .map(reward => RewardTransaction(address, TransactionAmount(reward)))
+                        .left
+                        .map(reason => RewardsException(reason))
+                  }
+                }.leftMap(_.asInstanceOf[RewardsError])
+              } yield (weightedDistribution.toSortedMap, SortedSet.from(transactions))
+          )
+
+      private[rewards] def getRewardsPerSnapshot(
+        snapshotOrdinal: SnapshotOrdinal,
+        timeTriggerInterval: FiniteDuration
+      ): Either[RewardsError, Amount] =
+        for {
+          snapshots <- getSnapshotsCountPerEpoch(timeTriggerInterval)
+          epoch <- getEpochNumber(snapshotOrdinal, snapshots)
+          reward <- getTotalRewardsPerEpoch(epoch)
+          // Note: We need rounding there to match validator rewards sheet rounding
+          rewardPerSnapshot <- NonNegLong
+            .from(Math.round(reward.value / snapshots.toDouble))
+            .left
+            .map(RewardsException)
+        } yield Amount(rewardPerSnapshot)
+
+      private[rewards] def getSnapshotsCountPerEpoch(
+        timeTriggerInterval: FiniteDuration
+      ): Either[RewardsError, PosLong] =
+        PosLong
+          .from((config.epochDurationInYears * 12 * 30 * 24 * 60).toLong / timeTriggerInterval.toMinutes)
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getTotalRewardsPerEpoch(epochNumber: PosInt): Either[RewardsError, Amount] =
+        NonNegLong
+          .from(
+            (config.baseEpochReward.value / Math.pow(2, epochNumber.toDouble - 1.0)).toLong
+          )
+          .map(Amount(_))
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getEpochNumber(
+        snapshotOrdinal: SnapshotOrdinal,
+        snapshotsPerEpoch: PosLong
+      ): Either[RewardsError, PosInt] =
+        snapshotOrdinal match {
+          case ordinal if ordinal == SnapshotOrdinal.MinValue => Right(1)
+          case ordinal =>
+            for (epoch <- 1 to config.epochs)
+              if (ordinal.value > (epoch - 1) * snapshotsPerEpoch && ordinal.value <= epoch * snapshotsPerEpoch)
+                return PosInt.from(epoch).left.map(RewardsException)
+            Left(LastEpochExceeded)
+
+        }
+
+      private[rewards] def createSharedDistribution(
+        snapshotHash: Hash,
+        snapshotReward: Amount,
+        addresses: NonEmptySet[Address]
+      ): F[Either[RewardsError, NonEmptyMap[Address, Amount]]] = {
+        val totalSpace = NonNegInt.unsafeFrom(addresses.length)
+        val (quotient, remainder) = snapshotReward.coerce.toLong /% totalSpace.toLong
```

Comment:
If you inline the `totalSpace` you get:
```
NonNegInt.unsafeFrom(addresses.length).toLong
```
which is doing nothing and is equivalent to just `address.length`.